src/kubernetes_cluster/proof/validation_rule.rs

```diff
@@ -0,0 +1,263 @@
+// Copyright 2022 VMware, Inc.
+// SPDX-License-Identifier: MIT
+#![allow(unused_imports)]
+use crate::external_api::spec::*;
+use crate::kubernetes_api_objects::{
+    api_method::*, common::*, dynamic::*, resource::*, stateful_set::*,
+};
+use crate::kubernetes_cluster::spec::{
+    cluster::*,
+    cluster_state_machine::Step,
+    controller::common::{ControllerActionInput, ControllerStep},
+    message::*,
+};
+use crate::reconciler::spec::reconciler::Reconciler;
+use crate::temporal_logic::{defs::*, rules::*};
+use vstd::prelude::*;
+
+verus! {
+
+impl <K: ResourceView, E: ExternalAPI, R: Reconciler<K, E>> Cluster<K, E, R> {
+
+pub open spec fn from_dynamic_preserves_spec() -> bool {
+    forall |d: DynamicObjectView|
+        #[trigger] K::from_dynamic_object(d).is_Ok()
+            ==> K::unmarshal_spec(d.spec).get_Ok_0() == K::from_dynamic_object(d).get_Ok_0().spec()
+}
+
+/// The relexitivity allows the metadata to be different.
+pub open spec fn is_reflexive_and_transitive() -> bool {
+    &&& forall |x: K, y: K|  x.spec() == y.spec() ==> #[trigger] K::transition_rule(x, y)
+    &&& forall |x: K, y: K, z: K| #![trigger K::transition_rule(x, y), K::transition_rule(y, z)]
+        K::transition_rule(x, y) && K::transition_rule(y, z) ==> K::transition_rule(x, z)
+}
+
+/// This spec and also this module are targeted at the relations of the three versions of custom resource object. We know that
+/// if cr is updated, the old and new object are subject to the transition rule (if any). Since scheduled_cr and triggering_cr
+/// are derived from the cr in etcd, they are either equal to or satisfy the transition rule with etcd cr.
+/// 
+/// When the transition rule is transitive, we can determine a linear order of the three custom resource objects.
+pub open spec fn etcd_and_scheduled_and_triggering_cr_in_correct_order(cr: K) -> StatePred<Self> {
```

Comment:
`correct_order` is a bit confusing. It seems to me you are proving the transition_rule is transitive in terms of space (etcd -> in schedule -> in reconcile), not just temporal.
May I suggest other names for this invariant, like `transition_rule_transits_from_etcd_to_controller`?